src/compiler/checker.ts

```diff
@@ -11904,6 +11910,16 @@ namespace ts {
             return true;
         }
 
+        function isAritySmaller(sourceType: Type, target: Expression) {
+            if (isFunctionExpressionOrArrowFunction(target) && isFunctionType(sourceType)) {
+                const sourceSignatures = getSignaturesOfType(sourceType, SignatureKind.Call);
+                const sourceLengths = sourceSignatures.map(sig => !sig.hasRestParameter ? sig.parameters.length : Number.MAX_VALUE);
+                return forEach(sourceLengths, len => len < target.parameters.length);

```

Comment:
A target with more parameters than the source might still be applicable if some of the target's parameters are optional. This doesn't appear to factor that in.